py/gc.c

```diff
@@ -222,6 +236,87 @@ void gc_add(void *start, void *end) {
     // Add this area to the linked list
     prev_area->next = area;
 }
+
+#if MICROPY_GC_SPLIT_HEAP_AUTO
+// Try to automatically add a heap area large enough to
+// fulfill 'failed_alloc'
+STATIC bool gc_try_add_heap(size_t failed_alloc) {
+    // 'needed' is the size of a heap large enough to hold failed_alloc, with
+    // the additional metadata overheads as calculated in gc_setup_area().
+    //
+    // Rather than reproduce that logic here, we save code size by estimating we
+    // need 103% of the failed allocation size. Implemented as (1+1/32) so the
+    // calculation can be add+shift rather than mul+div.
+    //
+    // This will be enough unless 'failed_alloc' is very small. For the case
+    // where 'failed_alloc' is very small, the new heap area size will be
+    // determined by the next stage in this function anyhow.
+    size_t needed = failed_alloc + (failed_alloc / 32);
+
+    size_t avail = gc_get_max_new_split();
+
+    DEBUG_printf("gc_try_add_heap failed_alloc " UINT_FMT ", "
+        "needed " UINT_FMT ", avail " UINT_FMT " bytes \n",
+        failed_alloc,
+        needed,
+        avail);
+
+    if (avail < needed || avail < 1024) {
```

Comment:
Could arguably delegate this `avail < 1024` failure case to the port by having `gc_get_max_new_split()` return zero in this case, but I think any other port that implements `MICROPY_GC_SPLIT_HEAP_AUTO` will have similar circumstances (lots of dynamic system heap usage), otherwise it's hard to see why the port would enable this feature.